src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -313,56 +320,197 @@ long heapAllocations()
+            private long count;
+
+
+            Builder(AbstractBounds<PartitionPosition> keyRange, PrimaryKey.Factory factory, int capacity)
+            {
+                this.min = factory.createTokenOnly(keyRange.left.getToken());
+                this.max = factory.createTokenOnly(keyRange.right.getToken());
+                this.keySets = new ArrayList<>(capacity);
             }
 
-            // skip entire partition keys if they don't overlap
-            if (!keyRange.right.isMinimum() && primaryKeys.first().partitionKey().compareTo(keyRange.right) > 0
-                || primaryKeys.last().partitionKey().compareTo(keyRange.left) < 0)
-                return;
+            public void add(PrimaryKeys primaryKeys)
+            {
+                if (primaryKeys.isEmpty())
+                    return;
+
+                int size = primaryKeys.size();
+                SortedSet<PrimaryKey> keys = primaryKeys.keys();
+                if (size == 1)
+                    keySets.add(keys.first());
+                else
+                    keySets.add(keys);
+
+                count += size;
+            }
 
-            for (PrimaryKey key : primaryKeys)
+            public int size()
             {
-                if (keyRange.contains(key.partitionKey()))
-                {
-                    mergedKeys.add(key);
+                return keySets.size();
+            }
 
-                    minimumKey = minimumKey == null ? key : key.compareTo(minimumKey) < 0 ? key : minimumKey;
-                    maximumKey = maximumKey == null ? key : key.compareTo(maximumKey) > 0 ? key : maximumKey;
-                }
+            public boolean isEmpty()
+            {
+                return keySets.isEmpty();
             }
+
+            public MergingRangeIterator build()
+            {
+                return new MergingRangeIterator(keySets, min, max, count);
+            }
+        }
+    }
+
+    static class SortedSetRangeIterator extends RangeIterator
+    {
+        private SortedSet<PrimaryKey> primaryKeySet;
+        private Iterator<PrimaryKey> iterator;
+        private PrimaryKey lastComputedKey;
+
+        public SortedSetRangeIterator(SortedSet<PrimaryKey> source)
+        {
+            super(source.first(), source.last(), source.size());
+            this.primaryKeySet = source;
+        }
+
+        @Override
+        protected PrimaryKey computeNext()
+        {
+            // Skip can be called multiple times in a row, so defer iterator creation until needed
+            if (iterator == null)
+                iterator = primaryKeySet.iterator();
+            lastComputedKey = iterator.hasNext() ? iterator.next() : endOfData();
+            return lastComputedKey;
+        }
+
+        @Override
+        protected void performSkipTo(PrimaryKey nextKey)
+        {
+            // Avoid going backwards
+            if (lastComputedKey != null && nextKey.compareTo(lastComputedKey) <= 0)
```

Comment:
Right, the RangeIterator#skipTo does not protect fully. Unfortunately I don't know which class skipped backwards, but I can easily check it by turning it into an assert.

Reply:
```
Caused by: java.lang.AssertionError: Attempt to skip backwards
	at org.apache.cassandra.index.sai.memory.TrieMemoryIndex$SortedSetRangeIterator.performSkipTo(TrieMemoryIndex.java:519)
	at org.apache.cassandra.index.sai.utils.RangeIterator.skipTo(RangeIterator.java:99)
	at org.apache.cassandra.index.sai.memory.FilteringKeyRangeIterator.performSkipTo(FilteringKeyRangeIterator.java:57)
	at org.apache.cassandra.index.sai.utils.RangeIterator.skipTo(RangeIterator.java:99)
	at org.apache.cassandra.index.sai.utils.LazyRangeIterator.performSkipTo(LazyRangeIterator.java:42)
	at org.apache.cassandra.index.sai.utils.RangeIterator.skipTo(RangeIterator.java:99)
	at org.apache.cassandra.index.sai.utils.RangeConcatIterator.performSkipTo(RangeConcatIterator.java:62)
	at org.apache.cassandra.index.sai.utils.RangeIterator.skipTo(RangeIterator.java:99)
	at org.apache.cassandra.index.sai.utils.RangeAntiJoinIterator.performSkipTo(RangeAntiJoinIterator.java:52)
	at org.apache.cassandra.index.sai.utils.RangeIterator.skipTo(RangeIterator.java:99)
	at org.apache.cassandra.index.sai.utils.TermIterator.performSkipTo(TermIterator.java:122)
	at org.apache.cassandra.index.sai.utils.RangeIterator.skipTo(RangeIterator.java:99)
	at org.apache.cassandra.index.sai.utils.RangeIntersectionIterator.nextOrNull(RangeIntersectionIterator.java:155)
	at org.apache.cassandra.index.sai.utils.RangeIntersectionIterator.computeNext(RangeIntersectionIterator.java:85)
	at org.apache.cassandra.index.sai.utils.RangeIntersectionIterator.computeNext(RangeIntersectionIterator.java:38)
```

Reply:
Fixed, it was `RangeAntiJoinIterator`

Reply:
We discussed this offline, and I understand the issue now. The problem was in this method, not in any other iterator. The `RangeAntiJoinIterator` triggered the issue by calling skipTo with a primary key less than the last returned primary key out of this iterator. That is within the contract though, and I think (though am not certain) that this could also happen with the `RangeIntersectionIterator`. The simple explanation is that we progressed forward in the `iterator` object but then we called `tailSet` on the `primaryKeySet` and we were able to go to an earlier key.